# Importing pandas into my application

In [603]:
import pandas as pd

# Loading my CSV file into a DataFrame

In [604]:
df = pd.read_csv('IOT-temp.csv')
df.head(10)

id  room_id/id      noted_date  temp  \
0  __export__.temp_log_196134_bd201015  Room Admin  8/12/2018 9:30    29   
1  __export__.temp_log_196131_7bca51bc  Room Admin  8/12/2018 9:30    29   
2  __export__.temp_log_196127_522915e3  Room Admin  8/12/2018 9:29    41   
3  __export__.temp_log_196128_be0919cf  Room Admin  8/12/2018 9:29    41   
4  __export__.temp_log_196126_d30b72fb  Room Admin  8/12/2018 9:29    31   
5  __export__.temp_log_196125_b0fa0b41  Room Admin  8/12/2018 9:29    31   
6  __export__.temp_log_196121_01544d45  Room Admin  8/12/2018 9:28    29   
7  __export__.temp_log_196122_f8b80a9f  Room Admin  8/12/2018 9:28    29   
8  __export__.temp_log_196111_6b7a0848  Room Admin  8/12/2018 9:26    29   
9  __export__.temp_log_196112_e134aebd  Room Admin  8/12/2018 9:26    29   

  out/in  
0     In  
1     In  
2    Out  
3    Out  
4     In  
5     In  
6     In  
7     In  
8     In  
9     In

In [605]:
df.shape

(97606, 5)

In [606]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 97606 entries, 0 to 97605
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   id          97606 non-null  object
 1   room_id/id  97606 non-null  object
 2   noted_date  97606 non-null  object
 3   temp        97606 non-null  int64 
 4   out/in      97606 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.7+ MB
None


# From the information displayed above we have four columns with datatype as object and only one column with datatype as integer.There are 97606 entries and 5 columns namely; id, room_id/id, noted_date, out/in.

# Renaming the Data columns

In [607]:
df = df.rename(columns = {
                            'id':'ID', 'noted_date':'DateTime', 'temp':'Temperature', 'out/in': 'Place'
                        }
              )

In [608]:
df.head()

ID  room_id/id        DateTime  \
0  __export__.temp_log_196134_bd201015  Room Admin  8/12/2018 9:30   
1  __export__.temp_log_196131_7bca51bc  Room Admin  8/12/2018 9:30   
2  __export__.temp_log_196127_522915e3  Room Admin  8/12/2018 9:29   
3  __export__.temp_log_196128_be0919cf  Room Admin  8/12/2018 9:29   
4  __export__.temp_log_196126_d30b72fb  Room Admin  8/12/2018 9:29   

   Temperature Place  
0           29    In  
1           29    In  
2           41   Out  
3           41   Out  
4           31    In

# Performing summary of the unique value count on each column

In [609]:
def get_df_summary(df):
    
    '''This function is used to summarise, especially unique value count and data type for variable'''
    
    unq_val_cnt_df = pd.DataFrame(df.nunique(), columns = ['unq_val_cnt'])
    unq_val_cnt_df.reset_index(inplace = True)
    unq_val_cnt_df.rename(columns = {'index':'variable'}, inplace = True)
    unq_val_cnt_df = unq_val_cnt_df.merge(df.dtypes.reset_index().rename(columns = {'index':'variable', 0:'dtype'}),
                                          on = 'variable')
    unq_val_cnt_df = unq_val_cnt_df.sort_values(by = 'unq_val_cnt', ascending = False)
    
    return unq_val_cnt_df

In [610]:
unq_val_cnt_df = get_df_summary(df)

In [611]:
unq_val_cnt_df

variable  unq_val_cnt   dtype
0           ID        97605  object
2     DateTime        27920  object
3  Temperature           31   int64
4        Place            2  object
1   room_id/id            1  object

# Dropping the "room_id/id" column since it has one value that won't be useful for analysis

In [612]:
df.drop(columns=["room_id/id"], inplace=True)

In [613]:
df.head()

ID        DateTime  Temperature Place
0  __export__.temp_log_196134_bd201015  8/12/2018 9:30           29    In
1  __export__.temp_log_196131_7bca51bc  8/12/2018 9:30           29    In
2  __export__.temp_log_196127_522915e3  8/12/2018 9:29           41   Out
3  __export__.temp_log_196128_be0919cf  8/12/2018 9:29           41   Out
4  __export__.temp_log_196126_d30b72fb  8/12/2018 9:29           31    In

# Observation: DateTime column has a mixture of '-' and '/' string as shown below. Relacing '-' to '/' in the DateTime Column

In [614]:
df.iloc[7609:7614]

ID          DateTime  Temperature Place
7609  __export__.temp_log_171463_f7832fdd    1/12/2018 0:04           37   Out
7610  __export__.temp_log_171457_5f37933a    1/12/2018 0:00           37   Out
7611  __export__.temp_log_171456_a6e6de9d    1/12/2018 0:00           37   Out
7612  __export__.temp_log_171452_ace9cea4  30-11-2018 23:58           37   Out
7613  __export__.temp_log_171453_55664bc9  30-11-2018 23:58           37   Out

# Based on above observation, replacing '-' to '/' in the DateTime Column

In [615]:
df['DateTime'] = df['DateTime'].str.replace('-', '/')

# Performing slicing between #7609 - #7613 to confirm successful replacement 

In [616]:
df.iloc[7609:7613]

ID          DateTime  Temperature Place
7609  __export__.temp_log_171463_f7832fdd    1/12/2018 0:04           37   Out
7610  __export__.temp_log_171457_5f37933a    1/12/2018 0:00           37   Out
7611  __export__.temp_log_171456_a6e6de9d    1/12/2018 0:00           37   Out
7612  __export__.temp_log_171452_ace9cea4  30/11/2018 23:58           37   Out

# Checking for Duplicates in the dataset. Returns True for every row that is a duplicate, otherwise False

In [617]:
print(df.duplicated())

0        False
1        False
2        False
3        False
4        False
         ...  
97601    False
97602    False
97603    False
97604    False
97605    False
Length: 97606, dtype: bool


In [618]:
print('No. of duplicate records in the data set: {}'.format(df.duplicated().sum()))

No. of duplicate records in the data set: 1


In [619]:
df[df.duplicated()]

ID        DateTime  Temperature Place
11  __export__.temp_log_196108_4a983c7e  8/12/2018 9:25           42   Out

# Using the returned duplicate records to compare other records in ID column to return similar duplicate record

In [620]:
df.loc[df['ID'] == '__export__.temp_log_196108_4a983c7e']

ID        DateTime  Temperature Place
10  __export__.temp_log_196108_4a983c7e  8/12/2018 9:25           42   Out
11  __export__.temp_log_196108_4a983c7e  8/12/2018 9:25           42   Out

# Dropping duplicate records returned above

In [621]:
df = df.drop_duplicates()

# Converting DateTime column into date-time

In [622]:
df["DateTime"] = pd.to_datetime(df["DateTime"], format='%d/%m/%Y %H:%M')

# Performing groupby on the dateframe using the DateTime column and sorting in ascending order to uncover duplicate records based on DateTime Values

In [623]:
df.groupby(['DateTime'])['DateTime'].count().sort_values(ascending = False).head()

DateTime
2018-09-12 03:09:00    65
2018-09-09 16:24:00    62
2018-09-09 17:31:00    53
2018-09-09 16:08:00    50
2018-09-11 19:42:00    50
Name: DateTime, dtype: int64

# using a record of the DateTime column to compare the entire DateTime column, and sort the output with the ID column to uncover patterns.

In [624]:
df.loc[df['DateTime'] == pd.to_datetime('12/9/2018 3:09', format = '%d/%m/%Y %H:%M')].sort_values(by = 'ID').head(10)

ID            DateTime  Temperature  \
61229  __export__.temp_log_101144_ff2f0b97 2018-09-12 03:09:00           29   
61258  __export__.temp_log_101502_172517d2 2018-09-12 03:09:00           29   
61255  __export__.temp_log_104868_a5e526b3 2018-09-12 03:09:00           28   
61231  __export__.temp_log_108845_062b2592 2018-09-12 03:09:00           28   
61272  __export__.temp_log_112303_fca608f4 2018-09-12 03:09:00           29   
61274  __export__.temp_log_112304_3fc1d307 2018-09-12 03:09:00           29   
61265  __export__.temp_log_113479_964d64b0 2018-09-12 03:09:00           29   
61270  __export__.temp_log_115449_e70330bd 2018-09-12 03:09:00           28   
61213  __export__.temp_log_116767_4dc92bc9 2018-09-12 03:09:00           28   
61268  __export__.temp_log_119813_6bb044a0 2018-09-12 03:09:00           28   

      Place  
61229   Out  
61258    In  
61255    In  
61231    In  
61272    In  
61274   Out  
61265    In  
61270   Out  
61213   Out  
61268    In

# from the above result output, taking a closer look at the values of the ID column, You will notice that it has a combination of numeric and alpha-numeric values that if properly analyzed can be used as primary key for our table.

# As seen in the above output, the numeric part of the ID column is on the 6th Index. I have to do further analysis to check if the 6th index will return the numeric part for all records in the ID column. If yes, the numeric part will be the suitable primary key for our table.

In [625]:
df['ID'].apply(lambda x : x.split('_')[6]).nunique() == df.shape[0]

True

In [626]:
def check_seventh_element(value):
    elements = value.split('_')
    if len(elements) >= 7:
        return parts[6]  # Return the 7th part
    else:
        return None  # Return None if the 7th part does not exist

# Applying the above function to the 'ID' column and creating a new column

In [627]:
df['ID_PK'] = df['ID'].apply(check_seventh_part)

# Performing sorting on the dataset using the ID_PK

In [628]:
df.loc[df['DateTime'] == pd.to_datetime('12/9/2018 3:09', format = '%d/%m/%Y %H:%M')].sort_values(by = 'ID_PK').head(10)

ID            DateTime  Temperature  \
61229  __export__.temp_log_101144_ff2f0b97 2018-09-12 03:09:00           29   
61258  __export__.temp_log_101502_172517d2 2018-09-12 03:09:00           29   
61255  __export__.temp_log_104868_a5e526b3 2018-09-12 03:09:00           28   
61231  __export__.temp_log_108845_062b2592 2018-09-12 03:09:00           28   
61272  __export__.temp_log_112303_fca608f4 2018-09-12 03:09:00           29   
61274  __export__.temp_log_112304_3fc1d307 2018-09-12 03:09:00           29   
61265  __export__.temp_log_113479_964d64b0 2018-09-12 03:09:00           29   
61270  __export__.temp_log_115449_e70330bd 2018-09-12 03:09:00           28   
61213  __export__.temp_log_116767_4dc92bc9 2018-09-12 03:09:00           28   
61268  __export__.temp_log_119813_6bb044a0 2018-09-12 03:09:00           28   

      Place   ID_PK  
61229   Out  101144  
61258    In  101502  
61255    In  104868  
61231    In  108845  
61272    In  112303  
61274   Out  112304  
61265    In  113479  
61270   Out  115449  
61213   Out  116767  
61268    In  119813

# Check the data type of the ID_PK column

In [629]:
print(df['ID_PK'].dtype)

object


# Converting ID_PK column from Object to numeric

In [630]:
df['ID_PK'] = pd.to_numeric(df['ID_PK'], errors='coerce')

In [631]:
df['ID_PK'].dtype

dtype('int64')

# Looking at the ID_PK column above, you will notice a wide gap between the IDs. Let's pick a range of values to further analyze the data and uncover patterns.

In [632]:
df.loc[df['ID_PK'].isin(range(101144, 101502))].sort_values(by='ID_PK')

ID            DateTime  Temperature  \
61229  __export__.temp_log_101144_ff2f0b97 2018-09-12 03:09:00           29   
71980  __export__.temp_log_101146_46d5babe 2018-09-11 10:20:00           35   
73063  __export__.temp_log_101150_232b02ab 2018-09-11 07:57:00           33   
70959  __export__.temp_log_101151_c651017f 2018-09-11 12:25:00           35   
72459  __export__.temp_log_101152_5edaffe0 2018-09-11 09:20:00           34   
...                                    ...                 ...          ...   
77892  __export__.temp_log_101492_2a517abc 2018-09-10 14:42:00           31   
58077  __export__.temp_log_101493_bbd2e151 2018-09-12 12:29:00           33   
14848  __export__.temp_log_101495_5eff48bf 2018-11-04 00:00:00           42   
15664  __export__.temp_log_101496_828059fc 2018-11-02 04:01:00           43   
63079  __export__.temp_log_101497_21f7596f 2018-09-12 00:49:00           27   

      Place   ID_PK  
61229   Out  101144  
71980    In  101146  
73063   Out  101150  
70959    In  101151  
72459   Out  101152  
...     ...     ...  
77892   Out  101492  
58077   Out  101493  
14848   Out  101495  
15664   Out  101496  
63079   Out  101497  

[189 rows x 5 columns]

In [636]:
df.loc[df['ID_PK'].isin(range(101492, 101495))].sort_values(by='ID_PK')

ID            DateTime  Temperature  \
77892  __export__.temp_log_101492_2a517abc 2018-09-10 14:42:00           31   
58077  __export__.temp_log_101493_bbd2e151 2018-09-12 12:29:00           33   

      Place   ID_PK  
77892   Out  101492  
58077   Out  101493

# Observations:
1. There is no record for ID_PK 101145
2. The ID_PK value is not serial. for instance, between ID_PK 101150 to 101152, records for ID_PK 101151 were taken at '2018-09-11 12:25:00' instead of at '2018-09-12 12:29:00', assuming ID_PK values are serial.
3. Based on the above findings, ID_PK can't be used to sort out our dataset but will remain a unique identifier.

# Selecting all the records in the ID column and replacing them with all records in the ID_PK column

In [638]:
df.loc[:, 'ID'] = df.loc[:, 'ID_PK']

In [640]:
df.head()

ID            DateTime  Temperature Place   ID_PK
0  196134 2018-12-08 09:30:00           29    In  196134
1  196131 2018-12-08 09:30:00           29    In  196131
2  196127 2018-12-08 09:29:00           41   Out  196127
3  196128 2018-12-08 09:29:00           41   Out  196128
4  196126 2018-12-08 09:29:00           31    In  196126

# Since I have successfully established a PK for the Dataset, and replaced it with the previous record of the ID column, I will have to drop the ID_PK column

In [642]:
df.drop(columns = 'ID_PK', inplace = True)

In [643]:
df.head()

ID            DateTime  Temperature Place
0  196134 2018-12-08 09:30:00           29    In
1  196131 2018-12-08 09:30:00           29    In
2  196127 2018-12-08 09:29:00           41   Out
3  196128 2018-12-08 09:29:00           41   Out
4  196126 2018-12-08 09:29:00           31    In

In [644]:
df.to_csv('IOT-temp_cleanup_pandas.csv')